pallets/gear-eth-bridge/src/lib.rs

```diff
@@ -0,0 +1,520 @@
+    #[pallet::generate_deposit(fn deposit_event)]
+    pub enum Event<T> {
+        /// Grandpa validator's keys set was hashed and set in storage at
+        /// first block of the last session in the era.
+        AuthoritySetHashChanged(H256),
+
+        /// Bridge got cleared on initialization of the second block in a new era.
+        BridgeCleared,
+
+        /// Optimistically, single-time called event defining that pallet
+        /// got initialized and started processing session changes,
+        /// as well as putting initial zeroed queue merkle root.
+        BridgeInitialized,
+
+        /// Bridge was paused and temporary doesn't process any incoming requests.
+        BridgePaused,
+
+        /// Bridge was unpaused and from now on processes any incoming requests.
+        BridgeUnpaused,
+
+        /// A new message was queued for bridging.
+        MessageQueued { message: EthMessage, hash: H256 },
+
+        /// Merkle root of the queue changed: new messages queued within the block.
+        QueueMerkleRootChanged(H256),
+    }
+
+    /// Pallet Gear Eth Bridge's error.
+    #[pallet::error]
+    #[cfg_attr(test, derive(Clone))]
+    pub enum Error<T> {
+        /// The error happens when bridge got called before
+        /// proper initialization after deployment.
+        BridgeIsNotYetInitialized,
+
+        /// The error happens when bridge got called when paused.
+        BridgeIsPaused,
+
+        /// The error happens when bridging message sent with too big payload.
+        MaxPayloadSizeExceeded,
+
+        /// The error happens when bridging queue capacity exceeded,
+        /// so message couldn't be sent.
+        QueueCapacityExceeded,
+
+        /// The error happens when bridging thorough builtin and message value
+        /// is inapplicable to operation or insufficient.
+        IncorrectValueApplied,
+    }
+
+    /// Lifecycle storage.
+    ///
+    /// Defines if pallet got initialized and focused on common session changes.
+    #[pallet::storage]
+    pub(crate) type Initialized<T> = StorageValue<_, bool, ValueQuery>;
+
+    /// Lifecycle storage.
+    ///
+    /// Defines if pallet is accepting any mutable requests. Governance-ruled.
+    #[pallet::storage]
+    pub(crate) type Paused<T> = StorageValue<_, bool, ValueQuery, ConstBool<true>>;
+
+    /// Primary storage.
+    ///
+    /// Keeps hash of queued validator keys for the next era.
+    ///
+    /// **Invariant**: Key exists in storage since first block of some era's last
+    /// session, until initialization of the second block of the next era.
+    #[pallet::storage]
+    pub(crate) type AuthoritySetHash<T> = StorageValue<_, H256>;
+
+    /// Primary storage.
+    ///
+    /// Keeps merkle root of the bridge's queued messages.
+    ///
+    /// **Invariant**: Key exists since pallet initialization. If queue is empty,
+    /// zeroed hash set in storage.
+    #[pallet::storage]
+    pub(crate) type QueueMerkleRoot<T> = StorageValue<_, H256>;
```

Comment:
Why not make it a `ValueQuery` kind then (not the default `OptionQuery`) thereby ensuring it always return some `OnEmpty` value (which can be defined as the zero hash)? This way there'll be no need to reset it to zero explicitly whenever the queue is empty.

Reply:
It's @mertwole request for the zero key existing in storage on empty